src/helpers/request.ts

```diff
@@ -0,0 +1,89 @@
+import { ParsedUrlQueryInput } from 'querystring';
+import { addQueryToUrl, appendPathnameToUrl } from 'url-transformers';
+import { compactDefined, flow } from './fp';
+import {
+  ABORTED,
+  ApiResponse,
+  handleFetchResponse,
+  HandleResponse,
+} from './response';
+import { OmitStrict } from './typescript';
+
+type BuildUrlParams = {
+  pathname: string;
+  query?: ParsedUrlQueryInput;
+};
+
+const buildUrl = ({ pathname, query = {} }: BuildUrlParams) =>
+  flow(appendPathnameToUrl(pathname), addQueryToUrl(compactDefined(query)));
+
+type FetchParams = Pick<RequestInit, 'method' | 'body' | 'headers'>;
+export type RequestParams = BuildUrlParams & FetchParams;
+
+/**
+ * No-op helper used to type-check the arguments.
+ */
+export const createRequestParams = (args: RequestParams) => args;
+
+type InitArguments = {
+  accessKey: string;
+  apiVersion?: string;
+  apiUrl?: string;
+} & OmitStrict<RequestInit, 'method' | 'body'>;
+
+type PromiseWithAbort<T> = Promise<T> & {
+  abort: AbortController['abort'];
+};
+
+type RequestGenerator<RequestArgs extends unknown[], ResponseType> = {
+  handleRequest: (...a: RequestArgs) => RequestParams;
+  handleResponse: HandleResponse<ResponseType>;
+};
+
+export const initMakeRequest = ({
+  accessKey,
+  apiVersion = 'v1',
+  apiUrl = 'https://api.unsplash.com',
+  headers: generalHeaders,
+  ...generalFetchOptions
+}: InitArguments) => <RequestArgs extends unknown[], ResponseType>({
+  handleResponse,
+  handleRequest,
+}: RequestGenerator<RequestArgs, ResponseType>) =>
+  flow(
+    handleRequest,
+    ({ pathname, query, method = 'GET', headers: endpointHeaders, body }) => {
+      const url = buildUrl({ pathname, query })(apiUrl);
+
+      const controller = new AbortController();
+      const signal = controller.signal;
+
+      const promise = fetch(url, {
+        method,
+        headers: {
+          ...generalHeaders,
+          ...endpointHeaders,
+          'Accept-Version': apiVersion,
+          Authorization: `Client-ID ${accessKey}`,
+        },
+        body,
+        signal,
+        ...generalFetchOptions,
+      })
+        .catch(err => {
+          if (err.name === 'AbortError') {
+            return ABORTED;
+          } else {
+            throw err;
+          }
```

Comment:
I'm curious why you decided to catch the `AbortError`. It might be more idiomatic/intuitive to consumers if we just mirror the behaviour of `fetch`, which is to throw the error. Of course, if someone chooses to wrap this with an observable, they can just omit the `AbortError` entirely, which is what `Rx.fromFetch` does when it wraps `fetch`: https://github.com/ReactiveX/rxjs/blob/6.5.5/src/internal/observable/dom/fetch.ts#L99

Reply:
That's a good point. This logic is definitely moving away from native spec behaviour. I didn't have `Rx.fromFetch` in mind when I did this, but it seems like I did something similar to them by treating the abort in a special way. My instinct was to avoid requiring consumers to handle the error because frankly, I find it odd that it's an Abort_Error_...there's no error, everything's fine 😅 . But there's not much we can do about that and we should mirror the spec like you said.

Our options:
- do nothing: let `AbortError`s get thrown as they usually would
- catch aborted reqs and return them in a special way like I have done (`result.type === 'aborted'`)
- do not integrate the `AbortController` at all: it would be fairly easy for those who want it to integrate it when they want to.

I'm starting to lean towards the 3rd one to keep this SDK more straightforward and with one less dependency (even though it's fairly small, users would have to add the `abort-controller` polyfill for certain browsers/node etc.). In fact, users who want to have a shared `AbortController` between multiple calls/endpoints will want to be able to override the `signal` config themselves. At that point, it feels best to drop it entirely 🤔 . WDYT?

Reply:
I removed abort controller: https://github.com/unsplash/unsplash-js/pull/143/commits/d26c3db8d540cfbfffcca0965ab005a0ad80cfaa

Reply:
I hadn't actually looked at the way you were using abort controllers/signals. The way I expected this would work is the user would create a controller themselves and pass the signal into the API function call. Is that how it works now?

Reply:
Yeah! That's how it'll work now 👍

Reply:
Yeah, although they can only pass the signal to the overall API object once:

```ts
const controller = new AbortController();
const signal = controller.signal;

const api = createApi({ accessKey: '', signal });
const photosPromise = api.photos.list().catch(err => {
  // handle abort error
});
const searchPromise = api.search.getPhotos({ query: 'dog' }).catch(err => {
  // handle abort error
});

controller.abort(); // will impact all calls
```

Allowing the user to provide a custom `signal` to an individual call/request would entail us updating the signature of every endpoint's `requestParams` to also include `signal`...not sure if that's worth the effort 🤔

Reply:
> Yeah, although they can only pass the signal to the overall API object once:

Ah, that could be quite limiting. Cancellation doesn't really make sense unless you can use it for individual/specific requests as opposed to all requests made by the client.

Maybe that's a good reason to pass the "dependencies" last instead of upfront (going back to this discussion https://github.com/unsplash/unsplash-web/pull/5900#discussion_r522091979).

Alternatively, perhaps we can find a way to merge the "dependencies" (access key and signal) into the request params, so the user only has to call one function:

```ts
const controller = new AbortController();
const signal = controller.signal;

const photosPromise = api.photos.list({ accessKey: '' }).catch(err => {
  // handle abort error
});
const searchPromise = api.search.getPhotos({ query: 'dog', accessKey: '', signal }).catch(err => {
  // handle abort error
});

controller.abort(); // will only abort the search request
```

Without this I don't think it would be possible for someone to wrap API functions in an `Observable` (one which calls `abort` on unsubscribe).

Reply:
You're right!

Having to provide the `accessKey` on every request seems highly unusual to me, so I'd rank this option last in terms of dev ergonomics.

I did some shuffling around so that users can provide an additional argument containing any `fetch` options on a per-call basis!